### PR TITLE
Add Token Caching Support For Managed Identity

### PR DESCRIPTION
**Pull request by g2vinay**
Integrates the Managed Identity Auth flow to use In Memory Token Caching

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 1]
```diff
@@ -30,23 +30,7 @@
...
-import com.microsoft.aad.msal4j.IClientCredential;
-import com.microsoft.aad.msal4j.InteractiveRequestParameters;
-import com.microsoft.aad.msal4j.MsalInteractionRequiredException;
-import com.microsoft.aad.msal4j.OnBehalfOfParameters;
-import com.microsoft.aad.msal4j.Prompt;
-import com.microsoft.aad.msal4j.PublicClientApplication;
-import com.microsoft.aad.msal4j.RefreshTokenParameters;
-import com.microsoft.aad.msal4j.SilentParameters;
-import com.microsoft.aad.msal4j.UserNamePasswordParameters;
+import com.microsoft.aad.msal4j.*;
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:33]: fix checkstyle.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 2]
```diff
@@ -424,6 +427,30 @@ public IdentityClientOptions setIdentityLogOptionsImpl(IdentityLogOptionsImpl id
         return this;
     }
 
+    public IdentityClientOptions setManagedIdentityType(ManagedIdentityType managedIdentityType) {
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:430]: add doc

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 3]
```diff
@@ -424,6 +427,30 @@ public IdentityClientOptions setIdentityLogOptionsImpl(IdentityLogOptionsImpl id
...
+        this.managedIdentityType = managedIdentityType;
+        return this;
+    }
+
+    public ManagedIdentityParameters getManagedIdentityParameters() {
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:435]: add doc

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 4]
```diff
@@ -424,6 +427,30 @@ public IdentityClientOptions setIdentityLogOptionsImpl(IdentityLogOptionsImpl id
...
+     *
+     * @param managedIdentityParameters the managed identity parameters to use for authentication.
+     * @return the identity log options.
+     */
+    public IdentityClientOptions setManagedIdentityParameters(ManagedIdentityParameters managedIdentityParameters) {
+        this.managedIdentityParameters = managedIdentityParameters;
+        return this;
+    }
+
+    public ManagedIdentityType getManagedIdentityType() {
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:450]: add doc

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityParameters.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package com.azure.identity.implementation;
+
+public class ManagedIdentityParameters {
+    private String identityEndpoint;
+    private String identityHeader;
+    private String msiEndpoint;
+    private String msiSecret;
+    private String identityServerThumbprint;
+
+    public ManagedIdentityParameters() { }
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityParameters.java:10]: add doc

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityType.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package com.azure.identity.implementation;
+
+public enum ManagedIdentityType {
```
Comment (g2vinay) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityType.java:3]: add doc for enum

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java [hunk 7]
```diff
@@ -83,13 +99,51 @@ public final class ManagedIdentityCredential implements TokenCredential {
...
+            case ARC:
+                return clientOptions
+                    .setManagedIdentityType(ManagedIdentityType.ARC)
+                    .setManagedIdentityParameters(new ManagedIdentityParameters()
+                        .setIdentityEndpoint(configuration.get(Configuration.PROPERTY_IDENTITY_ENDPOINT)));
+            case VM:
+                return clientOptions.setManagedIdentityType(ManagedIdentityType.VM);
+            case AKS:
+                return clientOptions.setManagedIdentityType(ManagedIdentityType.AKS);
+            default:
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java:142]: is `NONE` really a valid value? should this be an error?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 8]
```diff
@@ -277,6 +282,74 @@ private Mono<ConfidentialClientApplication> getConfidentialClientApplication() {
         });
     }
 
+    private Mono<ConfidentialClientApplication> getManagedIdentityConfidentialClient() {
+        return Mono.defer(() -> {
+            String authorityUrl = TRAILING_FORWARD_SLASHES.matcher(options.getAuthorityHost()).replaceAll("")
+                + "/" + tenantId;
+
+            // Temporarily pass in Dummy Client secret and Client ID. until MSal removes its requirements.
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:290]: do we have an issue tracking this we can link?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 9]
```diff
@@ -277,6 +282,74 @@ private Mono<ConfidentialClientApplication> getConfidentialClientApplication() {
...
+                return Mono.error(LOGGER.logExceptionAsWarning(new IllegalStateException(e)));
+            }
+
+            applicationBuilder.appTokenProvider(appTokenProviderParameters -> {
+                TokenRequestContext trc = new TokenRequestContext()
+                    .setScopes(new ArrayList<>(appTokenProviderParameters.scopes))
+                    .setClaims(appTokenProviderParameters.claims)
+                    .setTenantId(appTokenProviderParameters.tenantId);
+
+                Mono<AccessToken> accessTokenMono = getTokenFromTargetManagedIdentity(options.getManagedIdentityType(), trc);
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:307]: nit: appending Mono to the name doesn't seem to be our pattern.
```suggestion
                Mono<AccessToken> accessToken = getTokenFromTargetManagedIdentity(options.getManagedIdentityType(), trc);
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 10]
```diff
@@ -277,6 +282,74 @@ private Mono<ConfidentialClientApplication> getConfidentialClientApplication() {
...
+            if (options.getExecutorService() != null) {
+                applicationBuilder.executorService(options.getExecutorService());
+            }
+
+            ConfidentialClientApplication confidentialClientApplication = applicationBuilder.build();
+            return Mono.just(confidentialClientApplication);
+        });
+    }
+
+    private Mono<AccessToken> getTokenFromTargetManagedIdentity(ManagedIdentityType managedIdentityType, TokenRequestContext trc) {
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:335]: `managedIdentityType` is already on options so doesn't need to be passed here?

**Review by billwert — CHANGES_REQUESTED** [review]
1) Can we add better tests for the new codepath? I think the modification to the existing tests is just to ensure we hit the `getTokenFromTargetManagedIdentity` change.
2) In the places where a public credential type has been changed to call `authenticateWithManagedIdentityConfidentialClient` should we `private` the correponding `IdentityClient` method? (For example `authenticateToArcManagedIdentityEndpoint`

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java (hunk 10 above)
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:335]: nit: trc should be named `tokenRequestContext`.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityParameters.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
+package com.azure.identity.implementation;
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityParameters.java:1]: Add copyright header.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityType.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
...
+    VM, APP_SERVICE, SERVICE_FABRIC, ARC, AKS, NONE;
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityType.java:4]: nit: we should have one enum value per line.

Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ManagedIdentityType.java:4]: ManagedIdentityType `NONE` seems invalid. Should that be named `DEFAULT`?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 13]
```diff
@@ -277,6 +282,74 @@ private Mono<ConfidentialClientApplication> getConfidentialClientApplication() {
...
+
+            applicationBuilder.appTokenProvider(appTokenProviderParameters -> {
+                TokenRequestContext trc = new TokenRequestContext()
+                    .setScopes(new ArrayList<>(appTokenProviderParameters.scopes))
+                    .setClaims(appTokenProviderParameters.claims)
+                    .setTenantId(appTokenProviderParameters.tenantId);
+
+                Mono<AccessToken> accessTokenMono = getTokenFromTargetManagedIdentity(options.getManagedIdentityType(), trc);
+
+                return accessTokenMono.toFuture().thenApply(accessToken -> {
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:309]: Let's move `toFuture` to the end of the reactive stream.

`accessTokenMono.map(accessToken -> {...}).toFuture()`

**Review by srnagar — APPROVED** [review]

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java (hunk 8 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:290]: it is a topic for discussion at their next meeting, 
Depending on outcome an issue will be created by msal team.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java (hunk 7 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java:142]: Removed NONE

**Comment by g2vinay** [comment]
> 1. Can we add better tests for the new codepath? I think the modification to the existing tests is just to ensure we hit the `getTokenFromTargetManagedIdentity` change.
> 2. In the places where a public credential type has been changed to call `authenticateWithManagedIdentityConfidentialClient` should we `private` the correponding `IdentityClient` method? (For example `authenticateToArcManagedIdentityEndpoint`

For #1, We can add some Identity Client level helper method tests, which I will push in next commit, Msal callback flow is tested on msal end

For #2, It is an impl class, nothing that impacts users from API point of view, we can make it private, but won't create any user impact.

**Comment by billwert** [comment]
> For 2. It is an impl class, nothing that impacts users from API point of view, we can make it private, but won't create any user impact.

Yeah, I was thinking to be clear about our intent that these methods are not to be used from other creds in the future.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java (hunk 7 above)
Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java:142]: Hmm. I'm not sure that's entirely right either - now won't the default value be `VM`? We probably need a `NONE` or `DEFAULT` to indicate the unchosen value?

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java:142]: yeah in our code flow we default to VM, that's our default based on the logic we follow.
None or Unchosen are invalid states to have and are equivalent to the value being null.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 14]
```diff
@@ -288,31 +288,36 @@ private Mono<ConfidentialClientApplication> getManagedIdentityConfidentialClient
                 + "/" + tenantId;
 
             // Temporarily pass in Dummy Client secret and Client ID. until MSal removes its requirements.
-            IClientCredential credential = ClientCredentialFactory.createFromSecret("dummy-secret");
+            IClientCredential credential = ClientCredentialFactory
+                .createFromSecret(clientSecret != null ? clientSecret : "dummy-secret");
             ConfidentialClientApplication.Builder applicationBuilder =
-                ConfidentialClientApplication.builder(clientId == null ? IdentityConstants.DEVELOPER_SINGLE_SIGN_ON_ID
+                ConfidentialClientApplication.builder(clientId == null ? "SYSTEM-ASSIGNED-MANAGED-IDENTITY"
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:294]: should this be added to `IdentityConstants`?

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:294]: This is not a valid client id, and gets used only once, we can keep it local.
IdentityConstants holds valid variables currently.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 15]
```diff
@@ -1131,7 +1141,7 @@ public Mono<MsalToken> authenticateWithSharedTokenCache(TokenRequestContext requ
      * @param request the details of the token request
      * @return a Publisher that emits an AccessToken
      */
-    public Mono<AccessToken> authenticateToArcManagedIdentityEndpoint(String identityEndpoint,
+    Mono<AccessToken> authenticateToArcManagedIdentityEndpoint(String identityEndpoint,
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:1144]: default visibility is `package-private` which is not what I think we want here?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java (hunk 2 above)
Comment (billwert) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:429]: still missing doc

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java (hunk 15 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java]: made private.
package private helps to keep things accessible in tests, but tests does not use these directly currently.

**Review by billwert — APPROVED** [review]
:shipit:

**Pull request merged**